pkg/assembler/clients/helpers/bulk.go

```diff
@@ -180,110 +187,224 @@ func GetBulkAssembler(ctx context.Context, gqlclient graphql.Client) func([]asse
 				Software:     packageAndArtifactIDs,
 				Dependencies: isDependenciesIDs,
 				Occurrences:  isOccurrencesIDs,
-			}); err != nil {
+			}, collectedIDorPkgInputs, collectedIDorArtInputs); err != nil {
 				logger.Errorf("ingestHasSBOMs failed with error: %v", err)
 			}
 
 			logger.Infof("assembling VEX : %v", len(p.Vex))
-			if err := ingestVEXs(ctx, gqlclient, p.Vex); err != nil {
+			if err := ingestVEXs(ctx, gqlclient, p.Vex, collectedIDorPkgInputs, collectedIDorArtInputs, collectedIDorVulnInputs); err != nil {
 				logger.Errorf("ingestVEXs failed with error: %v", err)
 			}
 
 			logger.Infof("assembling HashEqual : %v", len(p.HashEqual))
-			if err := ingestHashEquals(ctx, gqlclient, p.HashEqual); err != nil {
+			if err := ingestHashEquals(ctx, gqlclient, p.HashEqual, collectedIDorArtInputs); err != nil {
 				logger.Errorf("ingestHashEquals failed with error: %v", err)
 			}
 
 			logger.Infof("assembling PkgEqual : %v", len(p.PkgEqual))
-			if err := ingestPkgEquals(ctx, gqlclient, p.PkgEqual); err != nil {
+			if err := ingestPkgEquals(ctx, gqlclient, p.PkgEqual, collectedIDorPkgInputs); err != nil {
 				logger.Errorf("ingestPkgEquals failed with error: %v", err)
 			}
 
 			logger.Infof("assembling CertifyLegal : %v", len(p.CertifyLegal))
-			if err := ingestCertifyLegals(ctx, gqlclient, p.CertifyLegal); err != nil {
+			if err := ingestCertifyLegals(ctx, gqlclient, p.CertifyLegal, collectedIDorPkgInputs, collectedIDorSrcInputs, collectedIDorLicenseInputs); err != nil {
 				logger.Errorf("ingestCertifyLegals failed with error: %v", err)
 			}
 		}
 		return nil
 	}
 }
 
-func ingestPackages(ctx context.Context, client graphql.Client, v []model.PkgInputSpec) ([]string, error) {
-	response, err := model.IngestPackages(ctx, client, v)
+// ingestPackages takes in the map of IDorPkgInput which contains the pkgInputSpec and outputs a map that contains the pkgIDs to be used for verb ingestion
+func ingestPackages(ctx context.Context, client graphql.Client, packageInputMap map[string]*model.IDorPkgInput) (map[string]*model.IDorPkgInput, error) {
+	var keys []string
+	var pkgInputs []model.IDorPkgInput
+	pkgInputs = make([]model.IDorPkgInput, 0)
```

Comment:
you can remove all these `slice = make...`. The append will create the slice.

Reply:
so this causes an error when the `packageInputMap` is nil. That is the reason to add all these in. It resulted in these errors:
```
IngestSources failed with error: returned error 422 Unprocessable Entity: {\"errors\":[{\"message\":\"cannot be null\",\"path\":[\"variable\",\"sources\"],\"extensions\":{\"code\":\"GRAPHQL_VALIDATION_FAILED\"}}],\"data\":null}",
```

Reply:
The resulting slice is nil if the append does not occur and this is not allowed by the schema.

Reply:
ah, got it.